New endpoints for template

sehkone opened the pull request:
GET/POST endpoints for the template, the predefined configuration for training.

atyoshioka commented on src/database/template.rs:
```diff
@@ -0,0 +1,155 @@
+    pub algorithm: Option<String>,
+    pub format: Option<Vec<Format>>,
+    pub dimension_default: Option<u32>,
+    pub dimensions: Option<Vec<u32>>,
```
Why we cannot use `Option<i64>` for `dimension_default` and `Option<Vec<i64>>` for `dimensions`? You simply convert `u32` to `i64` when inserting values without error handlings.

sehkone replied:
Good point. Even though frontend sends `u32`, no necessary for REview to insist `u32`, as you told. I changed all with `i64`.

atyoshioka commented on src/database/template.rs:
```diff
@@ -0,0 +1,155 @@
+        name: new_template.name.clone(),
+        event_type: new_template.event_type.clone(),
+        method: new_template.method.clone(),
+        algorithm: new_template.algorithm.clone(),
```
These `clone` are not needed.

sehkone replied:
Thanks for your meticulous review.

sehkone commented on src/database/template.rs:
```diff
@@ -0,0 +1,155 @@
+}
+
+#[derive(Debug, Queryable, Serialize)]
+struct TemplateTable {
```
I modified as you said.

sehkone commented on src/database/template.rs:
```diff
@@ -0,0 +1,155 @@
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub(crate) struct Format {
```
I removed it.

sehkone commented on src/database/template.rs:
```diff
@@ -0,0 +1,155 @@
+
+#[allow(clippy::module_name_repetitions)]
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub(crate) struct TemplateHttpTransfer {
```
I removed these unnecessary ones.

atyoshioka replied:
I still don't understand why `pub(crate)` is needed for this struct.

atyoshioka replied:
@sehkone I believe you missed this one.

sehkone replied:
Removed.

sehkone commented:
@atyoshioka,

I did what you reviewed in the above.

I added two hyperparameters in the `template` table and modified the GET endpoint to return not only all templates with no query but also one template with a `name`.

atyoshioka commented on migrations/2020-01-24-134220_create_template/up.sql:
```diff
@@ -0,0 +1,13 @@
+    min_token_length BIGINT,
+    eps DOUBLE PRECISION,
+    format JSONB,
+    dimension_default BIGINT, /* u32 should be used as BIGINT */
```
This comment is no longer needed.

sehkone replied:
Removed.

atyoshioka commented on src/database/template.rs:
```diff
@@ -0,0 +1,162 @@
+
+#[derive(Debug, Deserialize)]
+pub(crate) struct TemplateSelectQuery {
+    pub name: Option<String>,
```
Why `name` needs `pub`?

sehkone replied:
Removed.

atyoshioka commented on src/database/template.rs:
```diff
@@ -0,0 +1,162 @@
+    match query_result {
+        Ok(template) => {
+            let template: Vec<TemplateHttpTransfer> = template
+                .iter()
```
Use `into_iter()` instead to remove all unnecessary `clone` in `map`

sehkone replied:
Changed.

atyoshioka commented on src/database/template.rs:
```diff
@@ -0,0 +1,162 @@
+
+    let format = new_template
+        .format
+        .map(|f| serde_json::to_value(&f).unwrap());
```
Please use `expect` instead of `unwrap`. `unwrap` gives us an impression that this code is unsafe.

sehkone replied:
Corrected

atyoshioka commented on src/database/template.rs:
```diff
@@ -0,0 +1,162 @@
+    });
+
+    match insert_result {
+        Ok(category) => Ok(HttpResponse::Ok()
```
Any variable names would be fine but `category` is a bit strange.

sehkone replied:
Caused by copying of another block. Corrected.

sehkone commented:
@atyoshioka,

I did what you pointed out. Please, check this out.

atyoshioka commented on CHANGELOG.md:
```diff
@@ -22,6 +22,8 @@ project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 - `select` query to `GET /api/outlier`.
+- A new endpoint `GET /api/template` to fetch template(s) from template table.
+- A new endpoint `POST /api/template` to create a new template.
 
```
Please add `[Unreleased]` like before and move these two lines under it. Please do not forget to add the following link to the bottom links.

```
[Unreleased]: https://github.com/petabi/review/compare/0.7.0...master
```

sehkone commented:
@atyoshioka,

Please, review what I corrected.
